falcosecurity/libs
# fix(userspace): obtain a reliable process lineage from thread info

**What type of PR is this?**

/kind bug

/kind cleanup

/kind design

/kind feature

**Any specific area of the project related to this PR?**

/area libsinsp

/area tests

**Does this PR require a change in the driver versions?**

No

**What this PR does / why we need it**:

First of all, the size of this PR.

We have:
* +4308 lines of test only
* 74 new libsinsp tests
* 26 test files

So without tests we have "only"
* ~1800 lines
* 10 files touched


Considering that there are almost 400 lines of comments and the clone parsers refactor takes something like 700 lines we don't have a huge patch. Of course, I will split it if needed the issue is that libsinsp code is like glue so it is not easy to obtain working commits from a split. Let's say that if you want I can split the PR at this commit https://github.com/falcosecurity/libs/commit/707d6fb3736bcdb69917b9f1a8ee11a8d02a790e but the result will be a code that compiles with green tests, but not fully working and moreover the first part should be merged without tests, that is something I don't like too much. Let me know what you think :)


Now looking at the patch commit by commit
* there is a new struct called `thread_group_info`. This as the name suggest keeps track of  all the info regarding a thread group:
   * threads in the group
   * child_sub_reaper info
   * pid of the group
   * alive threads in the group
* remove flag `force` from `remove_thread` method, it is no more necessary, threads will be automatically purged when necessary
* add new utility methods in thread info, `is_dead()`, `is_invalid()` ,... and some methods to handle the thread group info
* assign a thread group info to each thread when it is created
* split the old clone parser into 2 separate parsers: one for the child and one for the caller
  * the clone syscall is one of the most complex in the Linux system and there are tons of flags and different behaviors, moreover, we have also clone events not generated by the sys_exit tracepoint but by other tracepoints like in ARM. So due to the huge number of possibilities mange the 2 cases of child and parent in the same parser seems like an overkill. Moreover, during this splitting process, I found some inconsistencies/issue...On top of my mind, we didn't consider some important flags like `CLONE_PARENT` and `CLONE_NEW_PID_NS`
* update some tests according to the new changes
* repair invalid threads when possible in execve, this means assigning a `ptid` and create a thread group info for them
* add a prctl parser to obtain child_sub_reaper info
* update the `remove_thread` logic using new concepts of thread group info and reaper
* adapt execve parser to delete all not leader threads when it is called. According to the manual when an execve is called all not leader threads should be removed and the main thread will take control of the execution flow.
   * See `resurrect_thread` method to discover a funny corner case with `stress-ng --exec`
* clean up all not used methods
* add a logic to clean expired threads and children when they are dead otherwise we will have an infinite list of weak_ptr increasing the memory footprint
* handle a very corner case in which the `init` process `tid=1` is not present in the system.
* minor fixes/cleanup
* add tons of tests for the new logic
* minor fixes/cleanup

Please note that right now there are 2 scap-files used for tests. they are not huge (30 MB, 60 MB) but probably best thing here is to push them somewhere and download them when necessary

**Which issue(s) this PR fixes**:

Fixes https://github.com/falcosecurity/libs/issues/1139

**Special notes for your reviewer**:

```release-note
fix(userspace): obtain a reliable process lineage from thread info
```
